### ISSUE TITLE
Speed tests and comparison to other Whisper versions

### ISSUE THREAD

**Issue by fquirin** [guillaumekln/faster-whisper#9]
Just wanted to say thanks for this great port of Whisper to CTranslate2.

I've done some tests and compared it to other ports like a TFlite version and a C++ version on Raspberry Pi 4. You can find the results [>here<](https://github.com/fquirin/speech-recognition-experiments#raspberry-pi-400---aarch64---debian-bullseye).

In conclusion it is as fast as the Tflite version, but smaller and has the better API right now 🙂 👍.

**Comment by guillaumekln** [comment]
Hey, thank you for this comparison! It's great to know that the implementation also works well on AArch64 systems.

Maybe you noticed that there were some issues with the English-only models (since you are using the multilingual model for CT2, but not the others). The issues are now fixed in the latest version. So if you want to run these models, you just need update the project and its dependencies.

**Comment by fquirin** [comment]
Exactly, I saw that there was something strange going on ^^.
I've installed the test version today, when did you update? :-)

**Comment by fquirin** [comment]
I've done a fresh install and still get some weird results with the tiny.en model. In my tests I use 2 files, a ~4s long file and a ~11s long file. With "tiny" I see this result:
```
Loading audio file: ../test-files/en_sh_lights_70pct_4s.wav
Result:
[0s -> 3s]  Set the lights in the living room to 70%.
Inference took 3.44s for 3.58s audio file.

Loading audio file: ../test-files/en_speech_jfk_11s.wav
Result:
[0s -> 11s]  And so my fellow Americans ask not what your country can do for you ask what you can do for your country.
Inference took 3.94s for 11.00s audio file.
```

And with the "tiny.en" model:
```
Loading audio file: ../test-files/en_sh_lights_70pct_4s.wav
Result:
[0s -> 4s]  Set the lights in the living room to 70%.
Inference took 3.14s for 3.58s audio file.

Loading audio file: ../test-files/en_speech_jfk_11s.wav
Result:
[0s -> 7s]  And so my fellow Americans ask not what your country can do for you
[7s -> 31s]  country.
Inference took 22.09s for 11.00s audio file.
```

Inference for the 4s audio is almost identical (slightly faster with tiny.en), but for the 11s audio its over **5 times slower**! 🤔

**Comment by nyadla-sys** [comment]
@guillaumekln 
Is the Whisper to CTranslate2 model available in full int8, with both the activations and weights expressed in int8

**Comment by guillaumekln** [comment]
@fquirin Thanks for the details. I can reproduce the issue in Docker container with ARM emulation (it works fine on my Intel CPU). The generation of the second segment is somehow incorrect and it triggers the "temperature fallback" which will try other predictions. This is why the result is both incorrect and slow. I will need to investigate to understand what is happening.

@nyadla-sys We use a mixed int8-float32 computation. Only the inputs and weights of linear layers are quantized to int8, the rest of the model is running in float32.

**Comment by nyadla-sys** [comment]
Is there a way for me to view the  whisper CTRanlate2 model and determine which layers are fully optimized and which ones are not?
if it is easy for you to list,please provide me the list of ops which are fully optimized

**Comment by guillaumekln** [comment]
I can use the original Python implementation as a reference.
* in all [linear layers](https://github.com/openai/whisper/blob/v20230124/whisper/model.py#L34-L38), `x` and `weight` are both int8
* when [computing logits](https://github.com/openai/whisper/blob/v20230124/whisper/model.py#L193), `x` and `self.token_embedding.weight` are both int8

**Comment by fquirin** [comment]
>I can reproduce the issue with the quantized tiny.en in a Docker container with ARM emulation

Oh nice, is this image available to download somewhere?

**Comment by guillaumekln** [comment]
I'm using [this approach](https://devopstales.github.io/linux/running_and_building_multi_arch_containers/) to test ARM binaries on my system.

Regarding the issue with the quantized tiny.en, I can also reproduce it with other configurations. For example when running on the GPU or enabling AVX instead of AVX2. So it looks like there are numerical issues when running this model in int8. I will check if there is anything we can do to improve the situation. Some models are harder to quantize than others.

On the other hand, there is no issue when the model is not quantized as far as I know.

**Comment by Mijawel** [comment]
Any idea why I might seem to be getting slower performance with this compared to regular whisper.

With the small model for a 20 second file, I'm getting about 7.5 seconds with ctranslate2 vs 4 seconds with regular whisper (this is on M1 Mac Pro)

On a GCP cloud intel chip (can't see exactly what but here are the log details), I am getting similarly slower times of almost double when using ctranslate2.

```
[ctranslate2] [thread 2] [info] CPU: GenuineIntel (SSE4.1=true, AVX=true, AVX2=true)
[ctranslate2] [thread 2] [info] - Selected ISA: AVX2
[ctranslate2] [thread 2] [info] - SGEMM backend: MKL (packed: false)
[ctranslate2] [thread 2] [info] - GEMM_S16 backend: MKL (packed: false)
[ctranslate2] [thread 2] [info] - GEMM_S8 backend: MKL (packed: false, u8s8 preferred: true)
[ctranslate2] [thread 2] [info] Loaded model whisper-small-ct2/ on device cpu:0
[ctranslate2] [thread 2] [info] - Binary version: 6
[ctranslate2] [thread 2] [info] - Model specification revision: 3
[ctranslate2] [thread 2] [info] - Selected compute type: int8
```

```
from faster_whisper import WhisperModel
import time

def main():
    model = WhisperModel("whisper-small-ct2/", device="cpu", compute_type="auto")

    startTime = time.time()
    print('Transcribing')
    segments, info = model.transcribe('chunk49_16khz.flac', beam_size=1)
    text = " ".join([segment.text for segment in segments])
    intermediateTime = time.time()
    print('Transcribed: %s' % (intermediateTime - startTime))
    return text

if __name__ == "__main__":
    main()
```

**Comment by guillaumekln** [comment]
Can you try running without quantization? Remove the argument `compute_type` here and also make sure to convert the model without an integer quantization.

**Comment by Mijawel** [comment]
Reconverted model with following command:

`ct2-transformers-converter --model openai/whisper-small --output_dir whisper-small-ct2-no-quant`

and ran without `compute_type` argument

Still getting same results between quantized and non quantized models (and both about half the speed of regular whisper)

**Comment by guillaumekln** [comment]
Is it possible for you to share the input audio file?

**Comment by Mijawel** [comment]
Yeh sure. Thanks so much for the help!

https://drive.google.com/file/d/1vFAHPswg9PQw-Lhq_jEy-2z5ivlWZp54/view?usp=sharing

**Comment by Mijawel** [comment]
Here's the example repo comparing the two:
https://github.com/Mijawel/faster_whisper_vs_whisper

(I only included the quantized model, as it was taking too long to upload the non quantized model, but I was receiving similar results with both)

**Comment by guillaumekln** [comment]
Thank you for sharing everything, that's very helpful!

There are several issues with your script:
* You can't compare `model.transcribe` and `model.decode` since they are not running the same thing. `transcribe` will actually call `decode` 2 times for this audio input.
* For faster-whisper you are not setting the language so it will run autodetection.
* You don't ensure the same number of CPU threads are used.

Here are the changes I propose to make the runs more comparable:

```diff
diff --git a/main.py b/main.py
index 9da4bdd..012b1d4 100644
--- a/main.py
+++ b/main.py
@@ -1,3 +1,6 @@
+import os
+os.environ["OMP_NUM_THREADS"] = "4"
+
 from faster_whisper import WhisperModel
 import time
 import whisper
@@ -8,7 +11,7 @@ def main():
 
     startTime = time.time()
     print('Transcribing with faster_whisper')
-    segments, info = model.transcribe('chunk49_16khz.flac', beam_size=1)
+    segments, info = model.transcribe('chunk49_16khz.flac', language="en", beam_size=1)
     text = " ".join([segment.text for segment in segments])
     intermediateTime = time.time()
     print('Transcribed: %s' % (intermediateTime - startTime))
@@ -19,16 +22,13 @@ def main():
 
     startTime = time.time()
     print('Transcribing with whisper')
-    audio = whisper.load_audio('chunk49_16khz.flac')
-    audio = whisper.pad_or_trim(audio)
-    mel = whisper.log_mel_spectrogram(audio).to(model.device)
-    options = whisper.DecodingOptions(fp16=False, language='en')
-    result = whisper.decode(model, mel, options)
+    result = model.transcribe('chunk49_16khz.flac', language="en", beam_size=1)
+    text = result["text"]
     intermediateTime = time.time()
     print('Transcribed: %s' % (intermediateTime - startTime))
-    print(result.text)
+    print(text)
 
     return text
```

With this change faster-whisper becomes 2x faster on my system:

```
Transcribing with faster_whisper
Transcribed: 4.429211616516113
 One is this Abel Bankew-Summer's Steckhauser paper, and now it's like one or two slides in the lecture today.  We didn't spend a lot of time on it in the lecture, but it's listed there as a supplemental reading along with my own paper.  I'll spend a little more time on my own paper.
Transcribing with whisper
Transcribed: 8.809089183807373
 One is this Abel, Thank You Summers Steckhouser paper, and now it's like one or two slides in the lecture today. We didn't spend a lot of time on it in the lecture, but it's listed there as a supplemental reading along with my own paper. I'll spend a little more time on my own paper.
```

**Comment by fquirin** [comment]
I've updated the [benchmark values with results from my brand new Orange Pi 5](https://github.com/fquirin/speech-recognition-experiments) SBC. Speed is incredible! 🚀 It's actually fast enough to deliver good user-experience in a voice-assistant scenario even without streaming support 🥳.

**Comment by Mijawel** [comment]
Thanks so much for that! Not seeing a 2x speed increase on my system but it is about 50% faster which is nice. (I'll try it on an intel cpu next)

Question is, why can't I run decode instead of transcribe with faster_whisper? (or whatever the equivalent is) if the output is the same?

(I'm only ever decoding files 30 seconds or less btw)

On another note. Every time I was starting up the container I noticed that it was re-downloading the tokenizer from the internet, which takes about 6 seconds. Which really slows down cold starts for anyone using this in a container form. (And I didn't see the same thing happening with whi

To solve, I used the following once, to get the file and package it with the source code:
`self.tokenizer.save("whisper_tokenizer.json")`

then replacing:
`self.tokenizer = tokenizers.Tokenizer.from_pretrained("openai/whisper-tiny")`

with:
`self.tokenizer = tokenizers.Tokenizer.from_file("whisper_tokenizer.json")`

**Comment by fquirin** [comment]
>On another note. Every time I was starting up the container I noticed that it was re-downloading the tokenizer from the internet, which takes about 6 seconds.

You can probably just commit the image (`sudo docker commit [running-container-name] [resulting-image-name]`) after you ran it once or add a volume for persistent data.

**Comment by Mijawel** [comment]
Ah thanks. BTW does the TFlite port of whisper have a 'maintained' repo like the CPP and CTranslate2 ports or is it just something you put together for the comparison? (because the times there are pretty impressive also)

**Comment by fquirin** [comment]
In the [README](https://github.com/fquirin/speech-recognition-experiments/blob/main/whisper-tflite/README.md) you'll always find the corresponding repositories 🙂.
Results are indeed pretty impressive, but it has some weird quirks that I don't see in CT2, for example transcription speed and results (for the German test) fluctuate. Not sure yet what's going on there. Tensorflow "fat" (pip install tensorflow) seems to have very good optimizations for ARM in general though (that are missing in default tflite_runtime for some reason).

**Comment by Mijawel** [comment]
Great thanks. will keep in mind

**Comment by guillaumekln** [comment]
> Question is, why can't I run decode instead of transcribe with faster_whisper? (or whatever the equivalent is) if the output is the same?

The equivalent is to call the [`generate`](https://github.com/guillaumekln/faster-whisper/blob/e2094b647497e18d1af4d5abfb2f9688d6166e35/faster_whisper/transcribe.py#L315-L322) method of the underlying model. But it is easier to always call `transcribe`.

**Comment by guillaumekln** [comment]
Thanks everyone for your contribution here!

I'm closing this issue to keep things organized. Feel free to open other issues to address or discuss specific points.
